keps/sig-architecture/3784-database-of-names/README.md

```diff
@@ -0,0 +1,449 @@
+    ValueTypeBool = ValueType("Bool")
+    ValueTypeJSONObject = ValueType("JSONObject")
+    ValueTypeJSONList = ValueType("JSONList")
+    // more...
+)
+```
+
+We can publish this schema as Go code or something else if there is a better or
+more standard notation.  We can validate all names against this schema.
+
+<<[UNRESOLVED enumerate resources ]>>
+Should we enumerate resources?  Good for validation, but a bit tedious.
+<<[/UNRESOLVED]>>
+
+#### Why YAML
+
+YAML is the lingua franca of Kubernetes.  We propose to use a subset of YAML
+which is closer to JSON and easier to avoid whitespace errors, while still
+allowing comments.
+
+#### Examples
+
+/by-name/io/kubernetes/names.yaml:
+```yaml
+[{
+  key: "kubernetes.io/hostname",
+  sig: "Node",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes" ],
+  description: "The name of the node.",
+}, {
+  key: "kubernetes.io/arch",
+  sig: "Node",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes" ],
+  description: "The architecture of the node.",
+}, {
+  key: "kubernetes.io/os",
+  sig: "Node",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes" ],
+  description: "The OS of the node.",
+}]
+```
+
+
+/by-name/io/kubernetes/topology/names.yaml:
+```yaml
+[{
+  key: "topology.kubernetes.io/zone",
+  sig: "Architecture",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes", core/persistentvolumes ],
+  description: "The availability zone of the resource.",
+  docsLink: "https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone"
+}, {
+  key: "topology.kubernetes.io/region",
+  sig: "Architecture",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes", core/persistentvolumes ],
+  description: "The availability region of the resource.",
+  docsLink: "https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesioregion"
+}]
+```
+
+<<[UNRESOLVED key collisions ]>>
+Should we have a YAML per key - easier lookups and collision avoidance, but
+lots of files.  Would maybe need Context to be a list.
+<<[/UNRESOLVED]>>
+
+### Risks and Mitigations
+
+Risk: Once we have a list, people will consider it authoritative and, even if
+we say otherwise, complete.  This does not absolve people of the need to "grep
+around" and make sure a name they want to use isn't quietly in use elsewehere.
+
+## Design Details
+
+See Proposal above.
+
+### Test Plan
+
+##### Prerequisite testing updates
+
+N/A
+
+##### Unit tests
+
+Tooling around this would be unit-tested.
+
+##### Integration tests
+
+N/A
+
+##### e2e tests
+
+N/A
+
+### Graduation Criteria
+
+CI must flag changes that do not conform to schema.
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+
+N/A
+
+## Implementation History
+
+Jan 26, 2023: First draft.
+
+## Drawbacks
+
+It's YET ANOTHER thing to update.
```

Comment:
Does this make it harder to localize the label / annotation definitions?

(I think: probably it does, but it's still do-able)

Reply:
Great insight - yes.  Perhaps we need to split the machine-data from the human-data?  E.g. one of the fields in the record is a pointer the human-friendly strings which are localized and can feed docs?  Ideally it would be one PR to update those - does that mean all the localized strings live in the same repo?  Or do we put EN in this DB and non-EN translations are a secondary DB which is used by docs?

Reply:
For the website, a common pattern is to generate [the relevant part of] the English docs using automation, and then have each localized version be generated and maintained by hand.

So the localized text would live only in k/website. I think that would work for this.